src/IECoreScene/SceneCache.cpp

```diff
@@ -728,8 +697,232 @@ class SceneCache::ReaderImplementation : public SceneCache::Implementation
 			return reader;
 		}
 
+		ConstPathMatcherDataPtr readSet( const Name &name ) const override
+		{
+			auto it = m_cachedSets.find( name );
+
+			if( it != m_cachedSets.end() )
+			{
+				return it->second;
+			}
+
+			ReaderImplementation *nc = const_cast<ReaderImplementation *>(this);
+			nc = nc->getRoot();
+			nc->convertTagsToSets();
+
+			IECore::PathMatcherDataPtr pathMatcherData = readLocalSet( name );
+
+			NameList childSetNames = readChildSetNames();
+
+			if( std::find( childSetNames.begin(), childSetNames.end(), name ) != childSetNames.end() )
+			{
+				NameList children;
+				childNames( children );
+
+				for( const auto &c : children )
+				{
+					ReaderImplementation *nc = const_cast<ReaderImplementation *>(this);
+					ConstPathMatcherDataPtr childSets = nc->child( c, SceneInterface::NullIfMissing )->readSet( name );
+					pathMatcherData->writable().addPaths( childSets->readable(), {c} );
+				}
+			}
+
+			m_cachedSets[name] = pathMatcherData;
+			return pathMatcherData;
+		}
+
+		NameList setNames() const override
+		{
+			ReaderImplementation *nc = const_cast<ReaderImplementation *>(this);
+			nc = nc->getRoot();
+			nc->convertTagsToSets();
+
+			IECore::CompoundDataBasePtr compoundDataBase = getSetCompound();
+
+			if( !compoundDataBase )
+			{
+				return NameList();
+			}
+
+			NameList setNames;
+			const auto &readable = compoundDataBase->readable();
+
+			for( auto it : readable )
+			{
+				setNames.push_back( it.first );
+			}
+
+			NameList childSetNames = readChildSetNames();
+
+			std::set<SceneInterface::Name> unique( setNames.begin(), setNames.end() );
+
+			for( const auto &childSetName : childSetNames )
+			{
+				unique.insert( childSetName );
+			}
+			return NameList( unique.begin(), unique.end() );
+		}
+
+		ReaderImplementation *getRoot()
+		{
+			if( m_parent )
+			{
+				return m_parent->getRoot();
+			}
+
+			return this;
+		}
+
+		void convertTagsToSets()
```

Comment:
Presumably this should be private? Little things like that would make it a lot easier to wrap your head around this sort of thing...